db/chat_message_storage.go

```diff
@@ -19,10 +25,355 @@ const (
+		if err != nil {
+			return err
+		}
+
+		// turn created at into bytes
+		// createdAtMsgID is the id used for the database
+		createdAtMsgID := make([]byte, 8)
+		binary.BigEndian.PutUint64(createdAtMsgID, uint64(msg.CreatedAt))
+
+		// make sure it is not taken and adjust the time indexed timestamp
+		tried := 0
+		for {
+			fetchedMsg := partnerBucket.Get(createdAtMsgID)
+			if fetchedMsg == nil || tried == 1000 {
+				break
+			}
+			tried++
+			binary.BigEndian.PutUint64(createdAtMsgID, uint64(msg.CreatedAt+1))
+		}
+
+		// set database id
+		msg.DatabaseID = int64(binary.BigEndian.Uint64(createdAtMsgID))
+
+		// marshal message
+		rawMessage, err := json.Marshal(msg)
+		if err != nil {
+			return err
+		}
+
+		// encrypt raw proto message
+		encryptedMessage, err := s.km.AESEncrypt(rawMessage)
+		if err != nil {
+			return err
+		}
+
+		// marshaled encrypted message
+		rawEncryptedMessage, err := encryptedMessage.Marshal()
+		if err != nil {
+			return err
+		}
+
+		// tell listeners that we persisted the message
+		tx.OnCommit(func() {
+			for _, listener := range s.postPersistListener {
+				go listener(MessagePersistedEvent{
+					Partner:     partner,
+					Message:     msg,
+					DBMessageID: int64(binary.BigEndian.Uint64(createdAtMsgID)),
+				})
+			}
+		})
+
+		return partnerBucket.Put(createdAtMsgID, rawEncryptedMessage)
+
+	})
+}
+
+// fetch all chat partners
+func (s *BoltChatMessageStorage) AllChats() ([]ed25519.PublicKey, error) {
+	chats := []ed25519.PublicKey{}
+	err := s.db.View(func(tx *bolt.Tx) error {
+
+		// all private chats
+		privateChats := tx.Bucket(privateChatBucketName)
+		if privateChats == nil {
+			return nil
+		}
+
+		return privateChats.ForEach(func(k, _ []byte) error {
+			if len(k) == 32 {
+				chats = append(chats, k)
+			}
+			return nil
+		})
+	})
+	return chats, err
+}
+
+func (s *BoltChatMessageStorage) Messages(partner ed25519.PublicKey, start int64, amount uint) (map[int64]Message, error) {
+
+	messages := map[int64]Message{}
+
+	err := s.db.View(func(tx *bolt.Tx) error {
+
+		// private chats
+		privChatsBucket := tx.Bucket(privateChatBucketName)
+		if privChatsBucket == nil {
+			return nil
+		}
+
+		// partner chat bucket
+		partnerBucket := privChatsBucket.Bucket(partner)
+		if partnerBucket == nil {
+			return nil
+		}
+
+		cursor := partnerBucket.Cursor()
+		var msgID int64
+		var rawMsg []byte
+
+		// jump to position
+		if start == 0 {
+			key, value := cursor.Last()
+			msgID = int64(binary.BigEndian.Uint64(key))
+			rawMsg = value
+		} else {
+			startBytes := make([]byte, 8)
+			binary.BigEndian.PutUint64(startBytes, uint64(start))
+			key, value := cursor.Seek(startBytes)
+			msgID = int64(binary.BigEndian.Uint64(key))
+			rawMsg = value
+		}
+
+		// unmarshal message
+		msg := Message{}
+		if err := json.Unmarshal(rawMsg, &msg); err != nil {
+			return err
+		}
+		messages[msgID] = msg
+
+		currentAmount := amount - 1
```

Comment:
We don't check anywhere if amount is already 0.

In the case it is, currentAmount would end up being 4294967295

And would pass the check inside the infinite for since we only check if currentAmount == 0. 

This would lead to unexpected behavior. 

Proposed solution below. 
```
var currentAmount uint
if amount > 0 {
     currentAmount = amount - 1
}
```

Reply:
fixed